internal/cloudsql/instance.go

```diff
@@ -0,0 +1,209 @@
+	ready chan struct{}
+}
+
+// Cancel prevents the instanceInfo from starting, if it hasn't already started. Returns true if timer
+// was stopped successfully, or false if it has already started.
+func (i *refreshResult) Cancel() bool {
+	return i.timer.Stop()
+}
+
+// Wait blocks until the refreshResult attempt is completed.
+func (i *refreshResult) Wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-i.ready:
+		return i.err
+	}
+}
+
+// Instance manages the information used to connect to the Cloud SQL instance by periodically calling
+// the Cloud SQL Admin API. It automatically refreshes the required information approximately 5 minutes
+// before the previous certificate expires (every 55 minutes).
+type Instance struct {
+	connName
+	client *sqladmin.Service
+	key    *rsa.PrivateKey
+
+	resultGuard sync.RWMutex
+	cur         *refreshResult
+	next        *refreshResult
+
+	// TODO: add a way to close
+}
+
+// NewInstance initializes a new Instance given an instance connection name
+func NewInstance(instance string, client *sqladmin.Service, key *rsa.PrivateKey) (*Instance, error) {
+	cn, err := parseConnName(instance)
+	if err != nil {
+		return nil, err
+	}
+	i := &Instance{
+		connName: cn,
+		client:   client,
+		key:      key,
+		cur:      nil,
+		next:     nil,
+	}
+	// For the inital refresh operation, set cur = next so that connection requests block
+	// until the first refresh is complete.
+	i.resultGuard.Lock()
+	i.cur = i.scheduleRefresh(0)
+	i.next = i.cur
+	i.resultGuard.Unlock()
+	return i, nil
+}
+
+// Connect returns a secure, authorized net.Conn to a Cloud SQL instance.
+func (i *Instance) Connect(ctx context.Context) (net.Conn, error) {
+	i.resultGuard.RLock()
+	res := i.cur
+	i.resultGuard.RUnlock()
+	err := res.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Add better ipType support, including an opt to specify.
+	addr := net.JoinHostPort(res.md.ipAddrs["PUBLIC"], "3307")
+	conn, err := proxy.Dial(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	if c, ok := conn.(*net.TCPConn); ok {
+		if err := c.SetKeepAlive(true); err != nil {
+			return nil, fmt.Errorf("failed to set keep-alive: %w", err)
+		}
+		if err := c.SetKeepAlivePeriod(defaultKeepAlive); err != nil {
+			return nil, fmt.Errorf("failed to set keep-alive period: %w", err)
+		}
+	}
+
+	tlsConn := tls.Client(conn, res.tlsCfg)
+	if err := tlsConn.Handshake(); err != nil {
+		tlsConn.Close()
+		return nil, fmt.Errorf("handshake failed: %w", err)
+	}
+	return tlsConn, err
+}
+
+// scheduleRefresh schedules a refresh operation to be triggered after a given duration. The returned resultRefresh
+// can be used to either Cancel or Wait for the operations result.
+func (i *Instance) scheduleRefresh(d time.Duration) *refreshResult {
+	res := &refreshResult{}
+	res.ready = make(chan struct{})
+	res.timer = time.AfterFunc(d, func() {
+		performRefresh(i, res, d)
```

Comment:
I think this function would read better if it returned a new `md`, `tlsConfig`, `err`, instead of mutating the passed in `res`. That would simplify the `performRefresh` signature and consolidate all mutation of `res` into one spot.